src/Server/HTTPHandler.cpp

```diff
@@ -322,10 +322,35 @@ bool HTTPHandler::authenticateUser(
     std::string user = request.get("X-ClickHouse-User", "");
     std::string password = request.get("X-ClickHouse-Key", "");
     std::string quota_key = request.get("X-ClickHouse-Quota", "");
+    std::string x509_user = request.get("X-ClickHouse-X509User", "");
```

Comment:
Adding another "username" field doesn't look good because we already have `X-ClickHouse-User`. If I understood correctly you only need a way to figure out if you need to extract a common name from a certificate or not. Maybe it's better to add a simple flag `X-ClickHouse-AuthWithSSLCertificate` (with possible values `no` and `common-name`)?

Reply:
the username is needed anyway,
otherwise there is no way to know which user to authenticate, which users settings to look against.

on the other hand, as you've mentioned, we need a switch to distinguish between different authentication types.
having non-empty X-ClickHouse-X509User header serves both purposes - we unambiguously know this is SSL certificate authemtication attempt and we know the username to authenticate against.

additionally, having two linked headers complicates the surrounding logic, as if the X-ClickHouse-User is specified we would have to ensure X-ClickHouse-AuthWithSSLCertificate is specified as well and if it has expected value, and if it's not - whether we should fallback (or assume) to X-ClickHouse-Key and X-ClickHouse-Quota and expect them to exist and have expected values.

here are the thoughts behind the X-ClickHouse-X509User header. what do you think?

Reply:
We usually use the same field for the same purpose. It doesn't seem to be necessary to make another `user name` field only to add another authentication type. And I don't think it's bad to have two or more linked headers, for example anyway we can have a quota specified in a separate header.

```
X-ClickHouse-User: john
X-ClickHouse-SSLCertificateAuth: by-common-name
X-ClickHouse-Quota: quotakey
```

Reply:
fixed